File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/CycleController.java
package com.panopset.blackjackEngine;

import java.io.StringWriter;

public class CycleController {

  private Cycle cycle;

  @Override
  public String toString() {
    StringWriter sw = new StringWriter();
    sw.append("cycle:" + cycle);
    return sw.toString();
  }

  public Cycle getCycle(BlackjackGameEngine g, Shoe s, Strategy strategy) {
    if (cycle == null) {
      cycle = new Cycle(g, s, strategy);
    }
    return cycle;
  }

  public synchronized void reset() {
    cycle = null;
  }

  public boolean isDealt() {
    if (cycle == null) {
      return false;
    }
    return cycle.isDealt();
  }

  public boolean isActive() {
    if (cycle == null) {
      return false;
    }
    return cycle.isActive();
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/CycleController.kt
package com.panopset.blackjackEngine

import java.io.StringWriter

class CycleController(val g: BlackjackGameEngine) {
    var cycle: Cycle? = null
    override fun toString(): String {
        val sw = StringWriter()
        sw.append("cycle:$cycle")
        return sw.toString()
    }

    fun getCycle(g: BlackjackGameEngine, strategy: Strategy): Cycle {
        if (cycle == null) {
            cycle = Cycle(g, strategy)
        }
        return cycle!!
    }

    @Synchronized
    fun reset() {
        cycle = null
    }

    fun cloneDealer(): HandDealer {
        return HandDealer(cycle?.dealer ?: HandDealer())
    }

    fun clonePlayers(): List<Player> {
        val clonedPlayers = ArrayList<Player>()
        if (cycle == null) {
            return clonedPlayers
        }
        for (player in cycle!!.players) {
            clonedPlayers.add(Player( player))
        }
        return clonedPlayers
    }

    val isDealt: Boolean
        get() = if (cycle == null) {
            false
        } else cycle!!.isDealt
    val isActive: Boolean
        get() = if (cycle == null) {
            false
        } else cycle!!.isActive
}
